lapack/gonum/dlatbs.go

```diff
@@ -0,0 +1,449 @@
+				// compute a solution to A*x = 0.
+				for i := range x[:n] {
+					x[i] = 0
+				}
+				x[j] = 1
+				xj = 1
+				scale = 0
+				xMax = 0
+			}
+
+			// Scale x if necessary to avoid overflow when adding a multiple of
+			// column j of A.
+			if xj > 1 {
+				rec := 1 / xj
+				if cnorm[j] > (bignum-xMax)*rec {
+					// Scale x by 1/(2*abs(x[j])).
+					rec *= 0.5
+					bi.Dscal(n, rec, x, 1)
+					scale *= rec
+				}
+			} else if xj*cnorm[j] > bignum-xMax {
+				// Scale x by 1/2.
+				bi.Dscal(n, 0.5, x, 1)
+				scale *= 0.5
+			}
+
+			if uplo == blas.Upper {
+				if j > 0 {
+					// Compute the update
+					//  x[max(0,j-kd):j] := x[max(0,j-kd):j] - x[j] * A[max(0,j-kd):j,j]
+					jlen := min(j, kd)
+					if jlen > 0 {
+						bi.Daxpy(jlen, -x[j]*tscal, ab[(j-jlen)*ldab+jlen:], kld, x[j-jlen:], 1)
+					}
+					i := bi.Idamax(j, x, 1)
+					xMax = math.Abs(x[i])
+				}
+			} else if j < n-1 {
+				// Compute the update
+				//  x[j+1:min(j+kd,n)] := x[j+1:min(j+kd,n)] - x[j] * A[j+1:min(j+kd,n),j]
+				jlen := min(kd, n-j-1)
+				if jlen > 0 {
+					bi.Daxpy(jlen, -x[j]*tscal, ab[(j+1)*ldab+kd-1:], kld, x[j+1:], 1)
+				}
+				i := j + 1 + bi.Idamax(n-j-1, x[j+1:], 1)
+				xMax = math.Abs(x[i])
+			}
+		}
+	} else {
+		// Solve A^T * x = b.
+		for j := jFirst; j != jLast; j += jInc {
+			// Compute x[j] = b[j] - sum A[k,j]*x[k].
+			//                       k!=j
+			xj := math.Abs(x[j])
+			tjjs := tscal
+			if diag == blas.NonUnit {
+				tjjs *= ab[j*ldab+maind]
+			}
+			tjj := math.Abs(tjjs)
+			rec := 1 / math.Max(1, xMax)
+			uscal := tscal
+			if cnorm[j] > (bignum-xj)*rec {
+				// If x[j] could overflow, scale x by 1/(2*xMax).
+				rec *= 0.5
+				if tjj > 1 {
+					// Divide by A[j,j] when scaling x if A[j,j] > 1.
+					rec = math.Min(1, rec*tjj)
+					uscal /= tjjs
+				}
+				if rec < 1 {
+					bi.Dscal(n, rec, x, 1)
+					scale *= rec
+					xMax *= rec
+				}
+			}
+
+			var sumj float64
+			if uscal == 1 {
+				// If the scaling needed for A in the dot product is 1, call
+				// Ddot to perform the dot product...
+				if uplo == blas.Upper {
+					jlen := min(j, kd)
+					if jlen > 0 {
+						sumj = bi.Ddot(jlen, ab[(j-jlen)*ldab+jlen:], kld, x[j-jlen:], 1)
+					}
+				} else {
+					jlen := min(n-j-1, kd)
+					if jlen > 0 {
+						sumj = bi.Ddot(jlen, ab[(j+1)*ldab+kd-1:], kld, x[j+1:], 1)
+					}
+				}
+			} else {
+				// ...otherwise, use in-line code for the dot product.
+				if uplo == blas.Upper {
+					jlen := min(j, kd)
+					for i := 0; i < jlen; i++ {
+						sumj += (ab[(j-jlen+i)*ldab+jlen-i] * uscal) * x[j-jlen+i]
+					}
+				} else {
+					jlen := min(n-j-1, kd)
+					for i := 0; i < jlen; i++ {
+						sumj += (ab[(j+1+i)*ldab+kd-1-i] * uscal) * x[j+i+1]
+					}
+				}
+			}
+
+			if uscal == tscal {
+				// Compute x[j] := ( x[j] - sumj ) / A[j,j]
+				// if 1/A[j,j] was not used to scale the dot product.
+				x[j] -= sumj
+				xj = math.Abs(x[j])
+				if diag == blas.NonUnit {
```

Comment:
I _think_ this should be `if diag == blas.NonUnit || tjjs != 1 {` but could you please also give it a thought to confirm?

Reply:
Could it even be only `if tjjs != 1 {`?

Reply:
If the required scaling is 1, the first path of the switch is taken, of which the conditional is not run and the elements of x are scaled by 1. In the case of a unit matrix, this still has not impact. For `tjjs != 1` then there is scaling. In the unit case, this looks to me like it will also scale the diagonal. If the user is storing information in the diagonal, this is probably not what they want.

So I think the `diag == blas.NonUnit` is required, but you could add `&& tjjs != 1` to elide the pointless scaling by 1.

Sorry to muddy the waters further.

Reply:
>  In the case of a unit matrix, this still has not impact. 

The problem that I see is that even in the case of a unit matrix there is the `tscal` factor which may cause `tjjs` to be non-unit. The reference executes the switch always for `diag == NonUnit` matrices _and_ for `Unit` matrices when `tscal != 1`. My code here executes the switch only for `NonUnit` matrices. So I still think it should have `|| tjjs != 1`. However, effectively what matters is not whether the matrix is Unit or NonUnit but rather what is the value that divides `x[j]` and therefore the simpler test `if tjjs != 1 {` should be enough?

> If the user is storing information in the diagonal, this is probably not what they want.
The diagonal elements in `ab` are accessed only for `NonUnit` matrices:
```
tjjs := tscal
if diag == blas.NonUnit {
	tjjs *= ab[j*ldab+maind]
}
```

Reply:
Or simply don't skip the switch at all. If `tjjs` is 1, so be it, there will be some negligible extra calculations. It's not being skipped in the `noTran` case either.

Reply:
Yes, sorry, I completely misread that.

I don't know. I will leave it up to you. Maybe @btracey wants to have a read?

Reply:
I've removed the skipping of the switch. I'm fairly confident that it is correct and the current tests pass ...